language/vm/vm-runtime/vm-runtime-types/src/values/values_impl.rs

```diff
@@ -0,0 +1,2223 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    loaded_data::{struct_def::StructDef, types::Type},
+    native_functions::dispatch::{native_gas, NativeResult},
+    native_structs::def::{NativeStructTag, NativeStructType},
+};
+use itertools::Itertools;
+use libra_types::{
+    account_address::{AccountAddress, ADDRESS_LENGTH},
+    byte_array::ByteArray,
+    language_storage::TypeTag,
+    vm_error::{sub_status::NFE_VECTOR_ERROR_BASE, StatusCode, VMStatus},
+};
+use std::{
+    cell::{Ref, RefCell, RefMut},
+    collections::VecDeque,
+    fmt::{self, Debug, Display},
+    mem::size_of,
+    ops::Add,
+    rc::Rc,
+};
+use vm::{
+    errors::*,
+    file_format::SignatureToken,
+    gas_schedule::{
+        words_in, AbstractMemorySize, CostTable, GasAlgebra, GasCarrier, NativeCostIndex,
+        CONST_SIZE, REFERENCE_SIZE, STRUCT_SIZE,
+    },
+};
+
+/***************************************************************************************
+ *
+ * Internal Types
+ *
+ *   Internal representation of the Move value calculus. These types are abstractions
+ *   over the concrete Move concepts and may carry additonal information that is not
+ *   defined by the language, but required by the implementation.
+ *
+ **************************************************************************************/
+
+/// Runtime representation of a Move value.
+#[derive(Debug)]
+enum ValueImpl {
+    Invalid,
+
+    U8(u8),
+    U64(u64),
+    U128(u128),
+    Bool(bool),
+    Address(AccountAddress),
+    ByteArray(ByteArray),
+
+    Container(Rc<RefCell<Container>>),
+
+    ContainerRef(ContainerRef),
+    IndexedRef(IndexedRef),
+}
+
+/// A container is a collection of values. It is used to represent data structures like a
+/// Move vector or struct.
+///
+/// There is one general container that can be used to store an array of any values, same
+/// type or not, and a few specialized flavors to offer compact memory layout for small
+/// primitive types.
+///
+/// Except when not owned by the VM stack, a container always lives inside an Rc<RefCell<>>,
+/// making it possible to be shared by references.
+#[derive(Debug)]
+enum Container {
+    General(Vec<ValueImpl>),
+    U8(Vec<u8>),
+    U64(Vec<u64>),
+    U128(Vec<u128>),
+    Bool(Vec<bool>),
+}
+
+/// A ContainerRef is a direct reference to a container, which could live either in the frame
+/// or in global storage. In the latter case, it also keeps a status flag indicating whether
+/// the container has possibly been modified.
+#[derive(Debug)]
+enum ContainerRef {
+    Local(Rc<RefCell<Container>>),
+    Global {
```

Comment:
It could be done that way, but I'd still prefer the current model for a reason different than the one you mentioned. 

In this model, `GlobalValue` is seen an owned value that lives in the `BTreeMap` in the `TransactionDataCache` and I'd like to have a clear sense of separation between an owned (but shared) value and a reference taken from it. I'm aware they share the same underlying representation, but having a `ContainerRef` owning a `GlobalValue` seems wrong to me. 

Another way to think of this, is to replace all `Rc` in references with `Weak` and you'll see them differ. We may actually do this if we want to ensure references don't outlive values.